src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs

```diff
@@ -0,0 +1,184 @@
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using DnsClient;
+using DnsClient.Protocol;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Clusters
+{
+    internal class DnsMonitor
+    {
+        #region static
+        private static string GetParentDomainName(string domainName)
+        {
+            var index = domainName.IndexOf('.');
+            return domainName.Substring(index + 1);
+        }
+        #endregion
+
+        // private fields
+        private readonly CancellationToken _cancellationToken;
+        private readonly IDnsMonitoringCluster _cluster;
+        private readonly LookupClient _lookupClient;
+        private readonly string _lookupDomainName;
+        private readonly string _parentDomainName;
+        private readonly string _query;
+        private bool _startReturnedNormally; // used for testing
+
+        // constructors
+        public DnsMonitor(IDnsMonitoringCluster cluster, string lookupDomainName, CancellationToken cancellationToken)
+        {
+            _cluster = Ensure.IsNotNull(cluster, nameof(cluster));
+            _lookupDomainName = Ensure.IsNotNull(lookupDomainName, nameof(lookupDomainName));
+            _cancellationToken = cancellationToken;
+            _lookupClient = new LookupClient();
+            _parentDomainName = GetParentDomainName(lookupDomainName);
+            _query = "_mongodb._tcp." + _lookupDomainName;
+        }
+
+        // public properties
+        public bool StartReturnedNormally => _startReturnedNormally;
+
+        // public methods
+        public void Start()
+        {
+            try
+            {
+                Monitor();
+            }
+            catch (OperationCanceledException)
+            {
+                // ignore OperationCanceledException
+            }
+            catch (Exception)
+            {
+                // TODO: log unexpected exceptions
+                throw;
+            }
+            _startReturnedNormally = true;
+        }
+
+        // private methods
+        private TimeSpan ComputeRescanDelay(List<SrvRecord> srvRecords)
+        {
+            var delay = TimeSpan.FromSeconds(60);
+
+            if (srvRecords.Count > 0)
+            {
+                var minTimeToLive = TimeSpan.FromSeconds(srvRecords.Select(s => s.InitialTimeToLive).Min());
+                if (minTimeToLive > delay)
+                {
+                    delay = minTimeToLive;
```

Comment:
I opened a SPEC ticket pointing out that if TTL has the DNS default of 24 hours we will only poll SRV once a day.